SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityTaskShuffler.java

```diff
@@ -0,0 +1,279 @@
+
+        if (resourceNotOverused || tooManyShufflingTasks) {
+          LOG.debug("Not shuffling any more tasks on slave {} ({} overage : {}%, shuffledOnHost: {}, totalShuffleCleanups: {})",
+              task.getTaskId().getSanitizedHost(),
+              slave.resource.resourceType,
+              slave.resource.overusage * 100,
+              shufflingTasksOnSlave,
+              shufflingTasksOnCluster
+          );
+          break;
+        }
+
+        String message = getShuffleMessage(slave, task, cpuUsage, memUsageBytes);
+        bounce(task, shuffleCleanupType, Optional.of(message));
+
+        cpuUsage -= task.getUsage().getCpusUsed();
+        memUsageBytes -= task.getUsage().getMemoryTotalBytes();
+
+        shufflingTasksOnSlave++;
+        shufflingTasksOnCluster++;
+        shufflingRequests.add(task.getTaskId().getRequestId());
+      }
+    }
+  }
+
+  private List<TaskIdWithUsage> getPrioritizedShuffleCandidates(OverusedSlave slave) {
+    // SingularityUsageHelper ensures that requests flagged as always ineligible for shuffling have been filtered out.
+    List<TaskIdWithUsage> out = slave.tasks;
+
+    if (slave.resource.resourceType == Type.CPU) {
+      out.sort((u1, u2) -> Double.compare(
+          getUtilizationScoreForCPUShuffle(u2),
+          getUtilizationScoreForCPUShuffle(u1)
+      ));
+    } else {
+      out.sort((u1, u2) -> Double.compare(
+          getUtilizationScoreForMemoryShuffle(u1),
+          getUtilizationScoreForMemoryShuffle(u2)
+      ));
+    }
+
+    return out;
+  }
+
+  private double getUtilizationScoreForCPUShuffle(TaskIdWithUsage task) {
+    return task.getUsage().getCpusUsed() / task.getRequestedResources().getCpus();
+  }
+
+  private double getUtilizationScoreForMemoryShuffle(TaskIdWithUsage task) {
+    double memoryUtilization = task.getUsage().getMemoryTotalBytes() / task.getRequestedResources().getMemoryMb();
+    double cpuUtilization = task.getUsage().getCpusUsed() / task.getRequestedResources().getCpus();
+
+    return (memoryUtilization + cpuUtilization) / 2;
+  }
+
+  private boolean isOverutilized(OverusedSlave slave, double cpuUsage, double memUsageBytes) {
+    if (slave.resource.resourceType == Type.CPU) {
+      return cpuUsage > slave.usage.getSystemCpusTotal();
+    } else {
+      return memUsageBytes > getTargetMemoryUtilizationForHost(slave.usage);
+    }
+  }
+
+  private boolean isShufflingTooManyTasks(long shuffledOnSlave, long shuffledOnCluster) {
+    return shuffledOnSlave > configuration.getMaxTasksToShufflePerHost()
+        || shuffledOnCluster > configuration.getMaxTasksToShuffleTotal();
+  }
+
+  private void bounce(TaskIdWithUsage task, TaskCleanupType cleanupType, Optional<String> message) {
+    taskManager.createTaskCleanup(new SingularityTaskCleanup(
+        Optional.empty(),
+        cleanupType,
+        System.currentTimeMillis(),
+        task.getTaskId(),
+        message,
+        Optional.of(UUID.randomUUID().toString()),
```

Comment:
Small nit, it would be good for the actionId here and in the SingularityPendingRequest to match. For this particular case it doesn't have any lasting effect on the actions themselves, but makes it easier to trace through the data for debugging if they are equal